pkg/operator/controller/controller.go

```diff
@@ -351,6 +342,68 @@ func (r *reconciler) ensureInternalRouterServiceForIngress(deployment *appsv1.De
 	return nil
 }
 
+// ensureRouterNamespaceAsset ensures that the router namespace exists and
+// matches the expected namespace.
+func (r *reconciler) ensureRouterNamespaceAsset() error {
+	ns, err := r.ManifestFactory.RouterNamespace()
+	if err != nil {
+		return fmt.Errorf("failed to build router namespace: %v", err)
+	}
+
+	namespaceMatcher := func(obj1, obj2 runtime.Object) bool {
+		ns1 := obj1.(*corev1.Namespace)
+		ns2 := obj2.(*corev1.Namespace)
+
+		if !reflect.DeepEqual(ns1.Labels, ns2.Labels) {
+			return false
+		}
+		if !reflect.DeepEqual(ns1.Annotations, ns2.Annotations) {
+			return false
+		}
+
+		return true
+	}
+
+	key := types.NamespacedName{Name: ns.Name}
+	if err := r.verifyRouterAssetExists(key, ns, namespaceMatcher); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// ensureRouterDeployment ensures that the router deployment exists and matches
+// the expected deployment.
+func (r *reconciler) ensureRouterDeployment(ci *ingressv1alpha1.ClusterIngress) (*appsv1.Deployment, error) {
+	deployment, err := r.ManifestFactory.RouterDeployment(ci)
+	if err != nil {
+		return nil, fmt.Errorf("failed to build router deployment: %v", err)
+	}
+
+	clusterIngressMatcher := func(obj1, obj2 runtime.Object) bool {
+		ci1 := obj1.(*ingressv1alpha1.ClusterIngress)
```

Comment:
Isn't `verifyRouterAssetExists` on L400 passing `Deployment`s?

Reply:
You're right, that's the wrong type! If tests weren't flaking, we'd be getting a nice run-time panic.

Reply:
line 400 calls `verifyRouterAssetExists` with `deployment` and that's passed back here w/ any current existing instance of a `deployment` object by that name aka calling `clusterIngressMatcher(current, expected)`. 

if you are referring to the failure in the tests : that looks to be from an old push -- because the line #s are way off + the variable name `current` does exist any more in that function. It looks like when the first force push happened, it ending up picking the first commit and not the second.  If not referring to the failure, ignore this paragraph!

Reply:
>line 400 calls verifyRouterAssetExists with deployment and that's passed back here w/ any current existing instance of a deployment object by that name aka calling clusterIngressMatcher(current, expected).

Yes, and then doesn't L384 try to cast the deployment to a *ingressv1alpha1.ClusterIngress?

>if you are referring to the failure in the tests 

I'm not